fast64_internal/f3d/f3d_texture_writer.py

```diff
@@ -0,0 +1,1358 @@
+        raise PluginError("Internal error, invalid use of getTextureNamesFromProp")
+    return getTextureNamesFromImage(texProp.tex, texProp.tex_format, parent)
+
+
+def checkDuplicateTextureName(parent: Union[FModel, FTexRect], name):
+    names = []
+    for info, texture in parent.textures.items():
+        names.append(texture.name)
+    while name in names:
+        name = name + "_copy"
+    return name
+
+
+def saveOrGetPaletteDefinition(
+    fMaterial: FMaterial,
+    parent: Union[FModel, FTexRect],
+    texProp: TextureProperty,
+    isPalRef: bool,
+    images: list[bpy.types.Image],
+    palBaseName: str,
+    palLen: int,
+) -> tuple[FPaletteKey, FImage]:
+
+    texFmt = texProp.tex_format
+    palFmt = texProp.ci_format
+    palFormat = texFormatOf[palFmt]
+    paletteKey = FPaletteKey(palFmt, images)
+
+    if isPalRef:
+        fPalette = FImage(texProp.pal_reference, None, None, 1, palLen, None)
+        return paletteKey, fPalette
+
+    # If palette already loaded, return that data.
+    fPalette = parent.getTextureAndHandleShared(paletteKey)
+    if fPalette is not None:
+        # print(f"Palette already exists")
+        return paletteKey, fPalette
+
+    paletteName, filename = getTextureNamesFromBasename(palBaseName, palFmt, parent, True)
+    fPalette = FImage(paletteName, palFormat, "G_IM_SIZ_16b", 1, palLen, filename)
+
+    parent.addTexture(paletteKey, fPalette, fMaterial)
+    return paletteKey, fPalette
+
+
+def saveOrGetTextureDefinition(
+    fMaterial: FMaterial,
+    parent: Union[FModel, FTexRect],
+    texProp: TextureProperty,
+    images: list[bpy.types.Image],
+    isLarge: bool,
+) -> tuple[FImageKey, FImage]:
+
+    image = texProp.tex
+    texFmt = texProp.tex_format
+    texFormat = texFormatOf[texFmt]
+    bitSize = texBitSizeF3D[texFmt]
+    imageKey = getImageKey(texProp, images)
+
+    if texProp.use_tex_reference:
+        width, height = texProp.tex_reference_size
+        fImage = FImage(texProp.tex_reference, None, None, width, height, None)
+        return imageKey, fImage
+
+    # If image already loaded, return that data.
+    fImage = parent.getTextureAndHandleShared(imageKey)
+    if fImage is not None:
+        # print(f"Image already exists")
+        return imageKey, fImage
+
+    imageName, filename = getTextureNamesFromProp(texProp, parent)
+    fImage = FImage(imageName, texFormat, bitSize, image.size[0], image.size[1], filename)
+    fImage.isLargeTexture = isLarge
+
+    parent.addTexture(imageKey, fImage, fMaterial)
+    return imageKey, fImage
+
+
+@dataclass
+class TexInfo:
+    # Main parameters
+    useTex: bool = False
+    isTexRef: bool = False
+    isTexCI: bool = False
+    texFormat: str = ""
+    palFormat: str = ""
+    imageDims: tuple[int, int] = (0, 0)
+    tmemSize: int = 0
+    errorMsg: str = ""
+
+    # Parameters from moreSetupFromModel
+    pal: Optional[list[int]] = None
+    palLen: int = 0
+    imUse: Optional[list[bpy.types.Image]] = None
```

Comment:
Renamed to `imDependencies` and `palDependencies`. These are lists of images which the outputted image or palette data depends on. For example, if you have two CI4 images sharing a 16-color palette:
- palDependencies is both of the input images
- imDependencies[0] is only the first input image, since its CIs are the same whether or not it is being rendered together with the other image or not
- imDependencies[1] is both of the input images, since its CIs will be shifted by the number of unique colors in the first input image

fast64 has to make these lists correctly for all cases of shared palettes, texture references, etc.

Reply:
ohhhhh `im` is image, i havent made that connection, though it makes sense considering `IM_RD`. imo using `image` would be clearer since it reflects the type used as well but this isnt a blocker, thanks for the updates!